### ISSUE TITLE
Support for Streams (Read/Write traits)?

### ISSUE THREAD

**Issue by julian-baumann** [mozilla/uniffi-rs#1485]
How do I expose streams? Or in rust terms, Read and Write traits.

Are there any plans to support streams as first class citizens? Since Swift and Kotlin both have native stream objects and as I read in the docs, one of the goals of this project is to translate the Rust api to these foreign languages as natively as possible.

In the meantime, I have also tried to implement it as best I can with current resources, but I have run into several problems, the main one being that `&mut` is not supported. Is there any way to work around this to make the following code work?

```rust
trait UniffiReadWrite {
    fn read_bytes(self: Arc<Self>, buffer: &Vec<u8>) -> Result<u64, ExternalIOError>;
    fn write_bytes(self: Arc<Self>, write_buffer: Vec<u8>) -> Result<u64, ExternalIOError>;
    fn flush_bytes(self: Arc<Self>) -> Result<(), ExternalIOError>;
}

impl UniffiReadWrite for EncryptedStream {
    fn read_bytes(self: Arc<Self>, buffer: &Vec<u8>) -> Result<u64, ExternalIOError> {
        return Ok(self.read(buffer.as_mut_slice())? as u64);
    }

    fn write_bytes(self: Arc<Self>, write_buffer: Vec<u8>) -> Result<u64, ExternalIOError> {
        return Ok(io::Write::write(Arc::get_mut(&mut self).unwrap(), write_buffer.as_slice())? as u64);
    }

    fn flush_bytes(self: Arc<Self>) -> Result<(), ExternalIOError> {
        return Ok(io::Write::flush(Arc::get_mut(&mut self).unwrap())?);
    }
}
```

with the corresponding UDL code:

```webidl
interface EncryptedStream {
    [Self=ByArc, Throws=ExternalIOError]
    u64 read_bytes([ByRef] sequence<u8> buffer);

    [Self=ByArc, Throws=ExternalIOError]
    u64 write_bytes(sequence<u8> buffer);
};
```

    error[E0596]: cannot borrow data in an `Arc` as mutable
      --> src/lib.rs:30:19
       |
    30 |         return Ok(self.read(buffer.as_mut_slice())? as u64);
       |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot borrow as mutable
       |
       = help: trait `DerefMut` is required to modify through a dereference, but it is not implemented for `Arc<EncryptedStream>`
    
    error[E0596]: cannot borrow `*buffer` as mutable, as it is behind a `&` reference
      --> src/lib.rs:30:29
       |
    30 |         return Ok(self.read(buffer.as_mut_slice())? as u64);
       |                             ^^^^^^^^^^^^^^^^^^^^^ `buffer` is a `&` reference, so the data it refers to cannot be borrowed as mutable
       |
    help: consider changing that to be a mutable reference
       |
    23 |     fn read_bytes(self: Arc<Self>, buffer: &mut Vec<u8>) -> Result<u64, ExternalIOError>;
       |                                            ~~~~~~~~~~~~
    
    error[E0596]: cannot borrow `self` as mutable, as it is not declared as mutable
      --> src/lib.rs:34:49
       |
    34 |         return Ok(io::Write::write(Arc::get_mut(&mut self).unwrap(), write_buffer.as_slice())? as u64);
       |                                                 ^^^^^^^^^ cannot borrow as mutable
       |
    help: consider changing this to be mutable
       |
    33 |     fn write_bytes(mut self: Arc<Self>, write_buffer: Vec<u8>) -> Result<u64, ExternalIOError> {
       |                    +++
    
    error[E0596]: cannot borrow `self` as mutable, as it is not declared as mutable
      --> src/lib.rs:38:49
       |
    38 |         return Ok(io::Write::flush(Arc::get_mut(&mut self).unwrap())?);
       |                                                 ^^^^^^^^^ cannot borrow as mutable
       |
    help: consider changing this to be mutable
       |
    37 |     fn flush_bytes(mut self: Arc<Self>) -> Result<(), ExternalIOError> {
       |                    +++
    

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/UNIFFI-243)
┆Components: UniFFI
┆Epic: Maintenance collector
┆Issue Number: UNIFFI-243
┆Sprint: Backlog

**Comment by mhammond** [comment]
Hey! Thanks for looking at UniFFI!

> How do I expose streams? Or in rust terms, Read and Write traits.
> Are there any plans to support streams as first class citizens?

There are no such plans.

> Since Swift and Kotlin both have native stream objects and as I read in the docs, one of the goals of this project is to translate the Rust api to these foreign languages as natively as possible.

I don't think it's fair to say that a goal of this project is to "translate *the* Rust api" where the definition of that is every trait and struct in `std` - if you can tell me what doc gave you that impression I'll put up a PR to try and avoid giving that impression.

> the main one being that &mut is not supported

As discussed [here](https://github.com/mozilla/uniffi-rs/blob/main/docs/adr/0004-only-threadsafe-interfaces.md), everything exposed via UniFFI must be `Send+Sync` and I think that alone is enough to prevent mutable references in most cases.

**Comment by julian-baumann** [comment]
Hey, sorry for my late reply. Was pretty busy with other stuff... 

"translate the Rust api" may just have been bad phrasing. What I actually meant was, that it seems to be a goal of this project to generate bindings that feel native to the language... So that you can use the generated swift code just as you would, if you would've written everyhing natively in swift. Since most (or all, I think) languages supported by uniffi support Streams, I was curious to find out whether there would be plans in introducing such an api.

But nevertheless, I still couldn't figure out, how to pass a mutable buffer from swift/kotlin into swift. I need to implement a read/write function, where I pass a buffer from swift/kotlin to the rust function. I would like to just use the pointer to that buffer, to avoid copying everything on every read/write call. 

Would you mind to give me a short example of how something like this could be done in uniffi? Or is there a better approach I should pursue?

**Comment by mhammond** [comment]
> it seems to be a goal of this project to generate bindings that feel native to the language

Yes, but with the additional complication that it must also be within the "rules" for Rust - handling a mutable reference out to foreign bindings without a way to manage the "lifetime" of that reference would be problematic.

> But nevertheless, I still couldn't figure out, how to pass a mutable buffer from swift/kotlin into swift. 

I'm afraid we don't have a story for that. We'd welcome a discussion though as it certainly is a reasonable use-case.

**Comment by julian-baumann** [comment]
In my scenario, I need to read from a network stream managed in Rust. Do you have any suggestions on how to accomplish this using Uniffi without relying on mutable buffers? Is it necessary to generate a new buffer with each read, and wouldn't this approach potentially result in significant memory usage?

**Comment by mhammond** [comment]
same as #1974, let's keep the newer one open.
